lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+	// ensure that threads waiting on them exit the compartment. We will
+	// empty the list later.
+	// TODO here we may want to check if the ring is well-formed
+	DebugErrorHandler::log(
+	  "Setting socket locks for destruction and destroying event groups.");
+	if (!sealedSockets.is_empty())
+	{
+		auto *cell = sealedSockets.first();
+		while (cell != &sealedSockets.sentinel)
+		{
+			struct SealedSocket *socket = SealedSocket::from_ring(cell);
+
+			FlagLockPriorityInherited *lock = &(socket->socketLock);
+			if (Capability{lock}.is_valid())
+			{
+				DebugErrorHandler::log("Destroying socket lock {}.", lock);
+				lock->upgrade_for_destruction();
+			}
+			else
+			{
+				DebugErrorHandler::log("Ignoring corrupted socket lock {}.",
+				                       lock);
+			}
+
+			FreeRTOS_Socket_t *s = socket->socket;
+			if (Capability{s}.is_valid() &&
+			    Capability{s->xEventGroup}.is_valid())
+			{
+				DebugErrorHandler::log("Destroying event group {}.",
+				                       s->xEventGroup);
+				eventgroup_destroy_force(MALLOC_CAPABILITY, s->xEventGroup);
+			}
+			else
+			{
+				DebugErrorHandler::log("Ignoring corrupted socket {}.", s);
+			}
+
+			cell = cell->cell_next();
+		}
+	}
+
+	DebugErrorHandler::log("Resetting the sealed sockets list.");
+	sealedSockets.reset();
+
+	// Upgrade the two critical section locks for destruction
+	// TODO document what this will do
+	DebugErrorHandler::log("Upgrading critical sections for destruction.");
+	flaglock_upgrade_for_destruction(&__CriticalSectionFlagLock.lock);
+	flaglock_upgrade_for_destruction(&__SuspendFlagLock.lock);
+
+	// Upgrade the message queue lock for destruction
+	DebugErrorHandler::log("Upgrading the message queue for destruction.");
+	// TODO document what this will do
+	auto *queueHandle = &xNetworkEventQueue->handle;
+	if (int err = queue_destroy(MALLOC_CAPABILITY, queueHandle); err != 0)
+	{
+		DebugErrorHandler::log(
+		  "Failed to upgrade the message queue for destruction (error {}).",
+		  err);
+	}
+
+	// Wait for all user threads to exit.
+	// TODO Here, we can experiment with `switcher_interrupt_thread` to get
+	// threads to die faster.
+	DebugErrorHandler::log("Waiting for all threads to exit.");
+	while (userThreadCount.load() != 0)
+	{
+		DebugErrorHandler::log("Waiting for {} user thread(s) to terminate.",
+		                       userThreadCount.load());
+
+		// Threads may also be waiting on the allocator in an
+		// out-of-memory situation. Do a `heap_free_all` to unblock
+		// them. We must do this in the loop body in case threads
+		// re-enter OOM multiple times.
+		//
+		// We will do another free at the end of the reset to ensure
+		// that everything is cleaned up in case threads allocate
+		// memory again before terminating.
+		free_compartment_memory();
+
+		Timeout t{1};
+		thread_sleep(&t);
+	}
+
+	// Wait for the IP thread to reset (unless this error handler is
+	// running from the IP thread).
+	if (isUserThread)
```

Comment:
I don't really like the amount of work that we do here.  I'd much rather that we did all of the cleanup on the TCP/IP thread, which has a known and predictable amount of stack space.  Running an error handler on a user thread means that the amount of stack space is unpredictable and may be very small, so we run a significant risk of double fault.

Ideally, all that this error handler would do would be to set a flag that we check in `ipFOREVER` and on new socket calls and then force unwind.  All other cleanup would run on the TCP/IP worker thread which, if it works once, we know definitely has enough stack space for all of the cleanup.

Reply:
The bits about cleaning up the locks we cannot avoid because they are necessary to unstuck the IP thread :(

(apart from the socket locks)

The bits at the end (after the `userThreadCount.load() == 0`) could be moved elsewhere. As we discussed separately this would be necessary to be more resilient against the situation where a thread enters the compartment with a very small stack.

Unfortunately this is not trivial to do, because the current setup process of the TCP/IP stack is from two threads: it's the user thread which initializes the TCP/IP stack through `network_start`, not the IP thread. Then the user thread unblocks the IP thread through the atomic entry guard. We need to rework that logic to do the initialization from the IP thread.

As discussed separately, let's keep a TODO list in an issue to track this an move with the PR as it is.

Reply:
To clarify this in the code, I added comments breaking down the code into bits which could be moved out of the error handler, and those which have to stay there.